sound/soc/sof/intel/hda-loader.c

```diff
@@ -335,6 +335,56 @@ int hda_dsp_cl_boot_firmware_iccmax(struct snd_sof_dev *sdev)
 	return ret;
 }
 
+static int hda_dsp_boot_imr(struct snd_sof_dev *sdev)
+{
+	struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
+	const struct sof_intel_dsp_desc *chip = hda->desc;
+	int ret, i;
+
+	/* init for booting wait */
+	init_waitqueue_head(&sdev->boot_wait);
+
+	/* power up & unstall/run core 0 to run the firmware */
+	ret = hda_dsp_enable_core(sdev, BIT(0));
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: dsp core start failed %d\n", ret);
+		return -EIO;
+	}
+
+	/* DSP is powered up, set all SSPs to slave mode */
+	for (i = 0; i < chip->ssp_count; i++) {
+		snd_sof_dsp_update_bits_unlocked(sdev, HDA_DSP_BAR,
+						 chip->ssp_base_offset
+						 + i * SSP_DEV_MEM_SIZE
+						 + SSP_SSC1_OFFSET,
+						 SSP_SET_SLAVE,
+						 SSP_SET_SLAVE);
+	}
+
+	/* enable IPC interrupts */
+	hda_dsp_ipc_int_enable(sdev);
+
+	/*
+	 * When a SoundWire link is in clock stop state, a Slave
+	 * device may trigger in-band wakes for events such as jack
+	 * insertion or acoustic event detection. This event will lead
+	 * to a WAKEEN interrupt, handled by the PCI device and routed
+	 * to PME if the PCI device is in D3. The resume function in
+	 * audio PCI driver will be invoked by ACPI for PME event and
+	 * initialize the device and process WAKEEN interrupt.
+	 *
+	 * The WAKEEN interrupt should be processed ASAP to prevent an
+	 * interrupt flood, otherwise other interrupts, such IPC,
+	 * cannot work normally.  The WAKEEN is handled after the ROM
+	 * is initialized successfully, which ensures power rails are
+	 * enabled before accessing the SoundWire SHIM registers
+	 */
+	if (!sdev->first_boot)
```

Comment:
what is the definition of 'first_boot' if we don't re-initialize the firmware?

Reply:
> what is the definition of 'first_boot' if we don't re-initialize the firmware?

We are only not re-downloading the firmware, the first/second/... FW boot concept are kept intact?

Reply:
my point was: be careful and check for all usages of 'first boot'. In the past we would download and authenticate the firmware at every resume, now only at first boot. It's a legitimate question to revisit all usages of first_boot to make sure we don't forget about something.

Reply:
> my point was: be careful and check for all usages of 'first boot'. In the past we would download and authenticate the firmware at every resume, now only at first boot. It's a legitimate question to revisit all usages of first_boot to make sure we don't forget about something.

Got you, let me list all of those usages below to relax us about that.

The first_boot is initialize to be 'true' and changed to 'false' once the probing succeed in core.c, all remain usages of 'first boot' is reading:

in hda-loader.c, 3 usages:
hda-loader.c:406, hda_sdw_process_waken() called for resuming boot, kept unchanged in my PR.
hda-loader.c:463, 487, hda_sdw_startup() called at _post_fw_run_() for first boot only, I think we don't need it for resuming case, even we don't re-download the FW, @bardliao @RanderWang @plbossart can you help to confirm this?

in loader.c, we basically use first_boot to decide if requesting firmware, creating debugfs and parsing fw_ready message is needed, we still don't need those (as previous) with the change here.

in pm.c:100, first_boot is used to check if sof_resume() should be performed, if the first_boot is true, it means the probing was not succeed so sof_resume() still no needed.

in trace.c:485, first_boot is used for creating trace related debugfs entries, we don't need create them for resuming boot, this is not changed.

```
$ grep first_boot sound/soc/sof/* -rn
sound/soc/sof/core.c:222:       sdev->first_boot = false;
sound/soc/sof/core.c:270:       sdev->first_boot = true;
sound/soc/sof/core.c:303:       sdev->first_boot = true;
sound/soc/sof/intel/hda-loader.c:406:   if (!sdev->first_boot)
sound/soc/sof/intel/hda-loader.c:463:   if (sdev->first_boot) {
sound/soc/sof/intel/hda-loader.c:487:   if (sdev->first_boot) {
sound/soc/sof/loader.c:66:      if (sdev->first_boot) {
sound/soc/sof/loader.c:190:     if (sdev->first_boot)
sound/soc/sof/loader.c:230:                     if (sdev->first_boot && elem->value)
sound/soc/sof/loader.c:495:     if (!sdev->first_boot)
sound/soc/sof/loader.c:795:     if (sdev->first_boot) {
sound/soc/sof/pm.c:100: if (sdev->first_boot)
sound/soc/sof/sof-priv.h:396:   bool first_boot;
sound/soc/sof/trace.c:485:      if (sdev->first_boot) {

```